server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java

```diff
@@ -178,79 +159,64 @@ public static Resolution ofOrdinal(int ord) {
-            this.ignoreMalformed = ignoreMalformed;
-            return builder;
-        }
+        private final Parameter<Boolean> index = Parameter.indexParam(m -> toType(m).indexed, true);
+        private final Parameter<Boolean> docValues = Parameter.docValuesParam(m -> toType(m).hasDocValues, true);
+        private final Parameter<Boolean> store = Parameter.storeParam(m -> toType(m).store, false);
 
-        protected Explicit<Boolean> ignoreMalformed(BuilderContext context) {
-            if (ignoreMalformed != null) {
-                return new Explicit<>(ignoreMalformed, true);
-            }
-            if (context.indexSettings() != null) {
-                return new Explicit<>(IGNORE_MALFORMED_SETTING.get(context.indexSettings()), false);
-            }
-            return Defaults.IGNORE_MALFORMED;
-        }
+        private final Parameter<Float> boost = Parameter.boostParam();
+        private final Parameter<Map<String, String>> meta = Parameter.metaParam();
 
-        public Builder locale(Locale locale) {
-            this.locale = locale;
-            return this;
-        }
+        private final Parameter<String> format
+            = Parameter.stringParam("format", false, m -> toType(m).format, DEFAULT_DATE_TIME_FORMATTER.pattern());
+        private final Parameter<Locale> locale = new Parameter<>("locale", false, Locale.ROOT,
+            (n, c, o) -> LocaleUtils.parse(o.toString()), m -> toType(m).locale);
 
-        public Builder nullValue(String nullValue) {
-            this.nullValue = nullValue;
-            return this;
-        }
+        private final Parameter<String> nullValue
+            = Parameter.stringParam("null_value", false, m -> toType(m).nullValueAsString, null);
+        private final Parameter<Boolean> ignoreMalformed;
 
-        public Locale locale() {
-            return locale;
-        }
-
-        public String format() {
-            return format.value();
-        }
-
-        public Builder format(String format) {
-            this.format = new Explicit<>(format, true);
-            return this;
-        }
+        private final Resolution resolution;
 
-        public Builder withResolution(Resolution resolution) {
+        public Builder(String name, Resolution resolution, DateFormatter dateFormatter, boolean ignoreMalformedByDefault) {
+            super(name);
             this.resolution = resolution;
-            return this;
+            this.ignoreMalformed
+                = Parameter.boolParam("ignore_malformed", true, m -> toType(m).ignoreMalformed, ignoreMalformedByDefault);
+            if (dateFormatter != null) {
+                this.format.setValue(dateFormatter.pattern());
+                this.locale.setValue(dateFormatter.locale());
+            }
         }
 
-        public boolean isFormatterSet() {
-            return format.explicit();
+        // for testing
+        public Builder(String name) {
```

Comment:
I'd prefer to drop these if we can manage it. Or to rename it to `public static Builder millisecondResolution(String name)` or something.